### PR TITLE
OS: Add support for Beagleboard

### PR DESCRIPTION
**Pull request by arrmo**
Includes Beagleboard and Beaglebone boards. Also includes temperature
sensor support (also in test data)

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [X] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)
- [X] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Comment by arrmo** [comment]
One question - should this be added to https://github.com/librenms/librenms/blob/master/doc/Extensions/Applications.md#raspberry-pi

As it's very similar (identical?) to the Raspberry Pi approach?

Thanks!

File: LibreNMS/OS/Beagleboard.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    OSDiscovery
+{
+    /**
+     * Retrieve basic information about the OS / device
+     */
+    public function discoverOS(): void
+    {
+        $device = $this->getDeviceModel();
+        $device->version = implode(' ', array_slice(explode(' ', trim(snmp_get($this->getDevice(), 'NET-SNMP-EXTEND-MIB::nsExtendOutputFull."distro"', '-Osqnv'), '"')), 1));
+        $device->hardware = trim(snmp_get($this->getDevice(), 'nsExtendOutputFull."hardware"', '-Osqnv'), '"');
```
Comment (Jellyfrog) [inline LibreNMS/OS/Beagleboard.php:39]: Can't you combine these snmp calls with snmp_get_multi()?

File: includes/discovery/sensors/temperature/beagleboard.inc.php [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+use LibreNMS\Config;
+
+$type = "beagleboardTemp";
+
+// $oid = 'NET-SNMP-EXTEND-MIB::nsExtendOutLine."beagleboard".';
+// Discovery is fine, but Polling seems to fail with string OID. So snmptranslate,
+//    snmptranslate -On NET-SNMP-EXTEND-MIB::nsExtendOutLine.\"beagleboard\"
+// to (with trailing ., to allow append in loop below),
+$oid = ".1.3.6.1.4.1.8072.1.3.2.4.1.2.11.98.101.97.103.108.101.98.111.97.114.100.";
+for ($sensor = 1; $volt <= 5; $sensor++) {
```
Comment (Jellyfrog) [inline includes/discovery/sensors/temperature/beagleboard.inc.php:17]: $volt?

File: includes/definitions/beagleboard.yaml [hunk 3]
```diff
@@ -0,0 +1,31 @@
+os: beagleboard
+type: server
+group: unix
+text: BeagleBoard
+icon: beagleboard
+ifXmcbc: true
+ifname: true
+mib_dir:
+    - supermicro
```
Comment (Jellyfrog) [inline includes/definitions/beagleboard.yaml:9]: This doesn't seem right?

File: includes/discovery/sensors/temperature/beagleboard.inc.php (hunk 2 above)
Comment (arrmo) [reply includes/discovery/sensors/temperature/beagleboard.inc.php:17]: Doh! 🤣. Copied this from Raspberry Pi, missed this one. Thanks!

Fixed, pushing updated files.

File: includes/definitions/beagleboard.yaml (hunk 3 above)
Comment (arrmo) [reply includes/definitions/beagleboard.yaml:9]: I admit, just left this the same as in linux, as that's what this device defaults to without these changes (i.e. if OS not added, it falls back to linux). Remove these lines?

Thanks!

File: LibreNMS/OS/Beagleboard.php (hunk 1 above)
Comment (arrmo) [reply LibreNMS/OS/Beagleboard.php:39]: I could (I think), but then still have to process them separately after. Was thinking this was clearer ... thoughts?

Comment (Jellyfrog) [reply LibreNMS/OS/Beagleboard.php:39]: It's more efficient with a multi query, see for example https://github.com/librenms/librenms/blob/master/LibreNMS/OS/Qnap.php

File: includes/definitions/beagleboard.yaml (hunk 3 above)
Comment (Jellyfrog) [reply includes/definitions/beagleboard.yaml:9]: Yes remove if not needed. I'm guessing beagleboard doesn't have its own mib

**Comment by Jellyfrog** [comment]
Tests fails also, https://travis-ci.com/github/librenms/librenms/jobs/381601311

**Comment by arrmo** [comment]
> Tests fails also, https://travis-ci.com/github/librenms/librenms/jobs/381601311

Yes, I saw that too, looked at the failure - it makes no sense to me 😞. Confirmed locally, and the result is as expected. Any thoughts?

Thanks!

File: includes/definitions/beagleboard.yaml (hunk 3 above)
Comment (arrmo) [reply includes/definitions/beagleboard.yaml:9]: Sorry, just to understand - you mean the full mib_dir section? I admit, not quite sure what is to be listed here, that's why I left it alone (i.e. as a copy from linux). Thanks!

Comment (Jellyfrog) [reply includes/definitions/beagleboard.yaml:9]: It's folders with mibs you want to include, to be able to use textual oids

Comment (arrmo) [reply includes/definitions/beagleboard.yaml]: Got it, thanks! Removed.

File: LibreNMS/OS/Beagleboard.php (hunk 1 above)
Comment (arrmo) [reply LibreNMS/OS/Beagleboard.php:39]: I can see that, from an snmp call perspective. But as I'm not using MIBs and names, the return value seems messed up with snmp_getnext_multi => the results become the array keys, not values? Data below, for example,
```
Array
(
    [BeagleBone AI] =>
    [18] =>
)
```

OK to leave it as is?

Thanks!

Comment (Jellyfrog) [reply LibreNMS/OS/Beagleboard.php:39]: It should work, see like https://github.com/librenms/librenms/blob/master/LibreNMS/OS/AviatWtm.php#L50

File: includes/discovery/sensors/temperature/beagleboard.inc.php [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+            $descr = "Core";
+            break;
+        case "4":
+            $descr = "DSP";
+            break;
+        case "5":
+            $descr = "IVA-HD";
+            break;
+    }
+    $value = trim(snmp_get($device, $oid . $sensor, '-Oqve'), '"');
```
Comment (Jellyfrog) [inline includes/discovery/sensors/temperature/beagleboard.inc.php:35]: Can we do a multi here also? Or get the entire $oid maybe?

File: LibreNMS/OS/Beagleboard.php (hunk 1 above)
Comment (arrmo) [reply LibreNMS/OS/Beagleboard.php:39]: Yep, this makes sense - but like you can see above, I'm getting "odd" results when not using named OID's (i.e. from an MIB). Rather, I'm using extend, so items like nsExtendOutputFull."hardware". It's not seeming to like that 😞.

I have tried the approach in both of the above - no joy.

File: includes/discovery/sensors/temperature/beagleboard.inc.php (hunk 4 above)
Comment (arrmo) [reply includes/discovery/sensors/temperature/beagleboard.inc.php:35]: I think I can get walk working here - give me a few 😆

Comment (arrmo) [reply includes/discovery/sensors/temperature/beagleboard.inc.php]: OK, pushed - I think this one is fixed ... agreed?

Thanks!

**Comment by arrmo** [comment]
> Tests fails also, https://travis-ci.com/github/librenms/librenms/jobs/381601311

I see the issue! Why is it assuming / using the os name (beagleboard) for the community? If I use public instead - it works. I admit, not sure where to set that.

File: LibreNMS/OS/Beagleboard.php (hunk 1 above)
Comment (arrmo) [reply LibreNMS/OS/Beagleboard.php]: Fixed! Had to use a slightly different function, and change the snmp options. But working now, Thanks!

File: LibreNMS/OS/Beagleboard.php [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+{
+    /**
+     * Retrieve basic information about the OS / device
+     */
+    public function discoverOS(): void
+    {
+        $device = $this->getDeviceModel();
+        $oids = ['nsExtendOutputFull."distro"', 'nsExtendOutputFull."hardware"'];
+        $info = snmp_get_multi($this->getDevice(), $oids);
+        $device->version = str_replace("BeagleBoard.org ", "", $info['"distro"'][nsExtendOutputFull]);
```
Comment (Jellyfrog) [inline LibreNMS/OS/Beagleboard.php:40]: missing quotes

File: LibreNMS/OS/Beagleboard.php [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        $device->hardware = $info['"hardware"'][nsExtendOutputFull];
```
Comment (Jellyfrog) [inline LibreNMS/OS/Beagleboard.php:41]: missing quotes

File: includes/discovery/sensors/temperature/beagleboard.inc.php [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+ * Temperature Sensors for BeagleBoard
+ * Requires snmp extend agent script from librenms-agent
+ */
+
+use Illuminate\Support\Str;
+use LibreNMS\Config;
+
+$type = "beagleboardTemp";
+
+// $oid = 'NET-SNMP-EXTEND-MIB::nsExtendOutLine."beagleboard"';
```
Comment (Jellyfrog) [inline includes/discovery/sensors/temperature/beagleboard.inc.php:12]: could you test this again? We hopefully fixed this

File: LibreNMS/OS/Beagleboard.php (hunk 5 above)
Comment (arrmo) [reply LibreNMS/OS/Beagleboard.php:40]: Fixed, thanks!

File: LibreNMS/OS/Beagleboard.php (hunk 6 above)
Comment (arrmo) [reply LibreNMS/OS/Beagleboard.php:41]: Fixed

File: includes/discovery/sensors/temperature/beagleboard.inc.php (hunk 7 above)
Comment (arrmo) [reply includes/discovery/sensors/temperature/beagleboard.inc.php:12]: Sure - updated, re-pushed. Let's see! 😄.

Was there an issue with the community string?

Comment (arrmo) [reply includes/discovery/sensors/temperature/beagleboard.inc.php:12]: No joy 😞.

File: LibreNMS/OS/Beagleboard.php [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+class Beagleboard extends OS implements
+    OSDiscovery
+{
+    /**
+     * Retrieve basic information about the OS / device
+     */
+    public function discoverOS(): void
+    {
+        $device = $this->getDeviceModel();
+        $oids = ['nsExtendOutputFull."distro"', 'nsExtendOutputFull."hardware"'];
```
Comment (Jellyfrog) [inline LibreNMS/OS/Beagleboard.php:38]:
```suggestion
        $oids = ['NET-SNMP-EXTEND-MIB::nsExtendOutputFull."distro"', 'NET-SNMP-EXTEND-MIB::nsExtendOutputFull."hardware"'];
```

Comment (arrmo) [reply LibreNMS/OS/Beagleboard.php]: OK, trying that - thanks!

**Review by Jellyfrog — APPROVED** [review]

**Comment by murrant** [comment]
This pull request has been mentioned on **LibreNMS Community**. There might be relevant details there:

https://community.librenms.org/t/v1-68-release-changelog-september-2020/13525/1

**Pull request merged**